node/impl/full/eth.go

```diff
@@ -2135,3 +2137,40 @@ func (g gasRewardSorter) Swap(i, j int) {
 func (g gasRewardSorter) Less(i, j int) bool {
 	return g[i].premium.Int.Cmp(g[j].premium.Int) == -1
 }
+
+func (a *EthModule) EthGetBlockReceipts(ctx context.Context, blockParam ethtypes.EthBlockNumberOrHash) ([]*api.EthTxReceipt, error) {
+	ts, err := getTipsetByEthBlockNumberOrHash(ctx, a.Chain, blockParam)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get tipset: %w", err)
+	}
+
+	messages, err := a.Chain.MessagesForTipset(ctx, ts)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get messages for tipset: %w", err)
+	}
+
+	receipts := make([]*api.EthTxReceipt, 0, len(messages))
+	for i, msg := range messages {
+		msgLookup, err := a.StateAPI.StateSearchMsg(ctx, ts.Key(), msg.Cid(), api.LookbackNoLimit, true)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to search for message: %w", err)
+		}
+		if msgLookup == nil {
+			continue
+		}
+
+		tx, err := newEthTxFromMessageLookup(ctx, msgLookup, i, a.Chain, a.StateAPI)
```

Comment:
I think a bunch of the cost in this call could be avoided by pulling up some of its functionality into this loop, or refactoring out a subset of that functionality that avoids loading the tipset again, loading the parent tipset to get the block hash and initialising the state tree. Those operations are all common across messages and you could do them once at the top of the loop and reuse inside, or use them to call a new function that's shared with `newEthTxFromMessageLookup`. `newEthTxFromSignedMessage` is the most important part of the call that you care about, most of the rest is common across all messages.

Something like this perhaps:

```diff
diff --git a/node/impl/full/eth_utils.go b/node/impl/full/eth_utils.go
index a9645ce9d..957b1bea0 100644
--- a/node/impl/full/eth_utils.go
+++ b/node/impl/full/eth_utils.go
@@ -643,19 +643,18 @@ func newEthTxFromMessageLookup(ctx context.Context, msgLookup *api.MsgLookup, tx
                }
        }
 
-       blkHash, err := ethtypes.EthHashFromCid(parentTsCid)
+       st, err := sa.StateManager.StateTree(ts.ParentState())
        if err != nil {
-               return ethtypes.EthTx{}, err
+               return ethtypes.EthTx{}, xerrors.Errorf("failed to load message state tree: %w", err)
        }
 
-       smsg, err := getSignedMessage(ctx, cs, msgLookup.Message)
-       if err != nil {
-               return ethtypes.EthTx{}, xerrors.Errorf("failed to get signed msg: %w", err)
-       }
+       return newEthTx(ctx, cs, st, parentTs.Height(), parentTsCid, msgLookup.Message, txIdx)
+}
 
-       st, err := sa.StateManager.StateTree(ts.ParentState())
+func newEthTx(ctx context.Context, cs *store.ChainStore, st *state.StateTree, blockHeight abi.ChainEpoch, parentTsCid cid.Cid, msgCid cid.Cid, txIdx int) (ethtypes.EthTx, error) {
+       smsg, err := getSignedMessage(ctx, cs, msgCid)
        if err != nil {
-               return ethtypes.EthTx{}, xerrors.Errorf("failed to load message state tree: %w", err)
+               return ethtypes.EthTx{}, xerrors.Errorf("failed to get signed msg: %w", err)
        }
 
        tx, err := newEthTxFromSignedMessage(smsg, st)
@@ -664,10 +663,15 @@ func newEthTxFromMessageLookup(ctx context.Context, msgLookup *api.MsgLookup, tx
        }
 
        var (
-               bn = ethtypes.EthUint64(parentTs.Height())
+               bn = ethtypes.EthUint64(blockHeight)
                ti = ethtypes.EthUint64(txIdx)
        )
 
+       blkHash, err := ethtypes.EthHashFromCid(parentTsCid)
+       if err != nil {
+               return ethtypes.EthTx{}, err
+       }
+
        tx.BlockHash = &blkHash
        tx.BlockNumber = &bn
        tx.TransactionIndex = &ti
```

And then you can use `newEthTx` from within this loop.